include/arch/arc/v2/irq.h

```diff
@@ -68,6 +69,82 @@ extern void z_irq_spurious(void *unused);
 })
 
 
+/**
+ * Configure a 'direct' static interrupt.
+ *
+ * When firq has no separate stack(CONFIG_ARC_FIRQ_STACK=N), it's not safe
+ * to call C ISR handlers because sp will be switched to bank1's sp which
+ * is undefined value.
+ * So for this case, the priority cannot be set to 0 but next level 1
+ *
+ * When firq has separate stack (CONFIG_ARC_FIRQ_STACK=y) but at the same
+ * time stack checking is enabled (CONFIG_ARC_STACK_CHECKING=y)
+ * the stack checking can raise stack check exception as sp is switched to
+ * firq's stack (bank1's sp). So for this case, the priority cannot be set
+ * to 0 but next level 1.
+ *
+ * Note that for the above cases, if application still wants to use firq by
+ * setting priority to 0. Application can call z_irq_priority_set again.
+ * Then it's left to application to handle the details of firq
+ *
+ * See include/irq.h for details.
+ * All arguments must be computable at build time.
+ */
+#define Z_ARCH_IRQ_DIRECT_CONNECT(irq_p, priority_p, isr_p, flags_p) \
+({ \
+	Z_ISR_DECLARE(irq_p, ISR_FLAG_DIRECT, isr_p, NULL); \
+	if (IS_ENABLED(CONFIG_ARC_FIRQ) && (priority_p == 0)) { \
+		if (!IS_ENABLED(CONFIG_ARC_FIRQ_STACK) || \
+		(IS_ENABLED(CONFIG_ARC_FIRQ_STACK) && \
+		IS_ENABLED(CONFIG_ARC_STACK_CHECKING))) { \
+			z_irq_priority_set(irq_p, 1, flags_p); \
+		} else { \
+			z_irq_priority_set(irq_p, 1, flags_p); \
+		} \
+	} else { \
+		z_irq_priority_set(irq_p, priority_p, flags_p); \
+	} \
+	irq_p; \
+})
+
+static inline void z_arch_isr_direct_header(void)
+{
+#ifdef CONFIG_TRACING
+	z_sys_trace_isr_enter();
+#endif
+}
+
+static inline void z_arch_isr_direct_footer(int maybe_swap)
+{
+	/* clear SW generated interrupt */
+	if (z_arc_v2_aux_reg_read(_ARC_V2_ICAUSE) ==
+	    z_arc_v2_aux_reg_read(_ARC_V2_AUX_IRQ_HINT)) {
+		z_arc_v2_aux_reg_write(_ARC_V2_AUX_IRQ_HINT, 0);
```

Comment:
If this is only for SW generated interrupts, why not leave this to the isr implementor?

Reply:
E.g. in zephyr's tests/kernel/gen_isr_table,  SW generated interrupts are used, but in the generic ISR, no ARC specific code is there to clear SW irq request.  Leaving the clear of SW int reqs to arch  will make the ISR more reuseable and more generic. This is also the way we did in _isr_demux.

Reply:
I think this current code is OK, by the time we get here we have already serviced the time-critical interrupt.